### DISCUSSION TITLE
Context

### DISCUSSION THREAD

**Discussion by rfjakob** [encryption-alliance/unified-vault-format#5] [General]
Hi, nice that you pick the common-format idea up again.

FYI I don't have a lot of time on my hands at the moment to work on gocryptfs, so it seems unlikely that I can make big changes anytime soon.

That said, I looked trough https://docs.cryptomator.org/en/latest/security/architecture/ and I like how directory flattening works. Any trouble with that scheme apart from sync apps getting confused (as mentioned)?

What's your motivation for changing the format for Cryptomator? To get something faster than AES-CTR + HMAC?

**Comment by rfjakob** [comment]
About content encryption, have you thought about chacha20-poly1305?

I have recently added it to gocryptfs, and I'm considering making it the default because it performs so well even without AES acceleration ( for example: https://github.com/rfjakob/gocryptfs/wiki/CPU-Benchmarks#raspberry-pi-4-model-b-bcm2835-4-x-cortex-a72--15ghz )

Then, a question, why 32kiB block size?

**Reply by overheadhunter** [reply]
> About content encryption, have you thought about chacha20-poly1305?
> 
> I have recently added it to gocryptfs, and I'm considering making it the default because it performs so well even without AES acceleration ( for example: https://github.com/rfjakob/gocryptfs/wiki/CPU-Benchmarks#raspberry-pi-4-model-b-bcm2835-4-x-cortex-a72--15ghz )

I'm certainly open for adding different ciphers. Some Cryptomator users are asking for this as well.

What I would like use to do is define a set of ciphers / cipher modes / block sizes. Some of them are required "MUST", some recommended "SHOULD" to be supported by applications that follow our standard.

We should be careful though, as this may lead to incompatibilities, if an application supports only one of the multiple methods. Therefore, we should start with a very small set.

> Then, a question, why 32kiB block size?

I guess you ask because of this line:

https://github.com/cryptomator/new-encryption-format/blob/89470cb0524d1826c728d429aa675857ff91d531/file%20content%20encryption/README.md?plain=1#L21

It is just an example to get the document started. 4k is fine as well. Cryptomator uses 32k blocks, as it hit a performance sweet spot. In fact larger multimedia files perform better when sequentially en-/decrypting in even larger blocks, as we don't need to re-initialize the cipher all the time. But it slows down random write, as the whole block needs to be re-written to avoid nonce-reuse.

I can even imagine using different encryption methods for different files within the same vault. The used method could be stored in the file header. An implementation of our standard could then decide to apply different encryption methods for different files (e.g. by sniffing the content type from the first few bytes).

**Reply by rfjakob** [reply]
Yes, I also think the "MUST" set should be very small. Ideally, there's only option for everything (i.e. no "SHOULD" at all).

**Reply by rfjakob** [reply]
PS: I don't know a lot about Java, but looks like ChaCha20-Poly1305 is in the stdlib: https://openjdk.org/jeps/329 . But AES-GCM is fine for me as well. It's still the default in gocryptfs, after all, and it's fine for 95% of the usecases.

**Reply by overheadhunter** [reply]
We need to look into mobile devices as well and should keep the Web Crypto API in mind. AES-GCM has certainly the broadest support but ChaCha isn't that exotic any more either. Offering alternatives doesn't hurt.

**Comment by overheadhunter** [comment]
> FYI I don't have a lot of time on my hands at the moment to work on gocryptfs, so it seems unlikely that I can make big changes anytime soon.

No worries, it'll take us some time to switch as well, but you can't start early enough to define a common standard.

> That said, I looked trough https://docs.cryptomator.org/en/latest/security/architecture/ and I like how directory flattening works. Any trouble with that scheme apart from sync apps getting confused (as mentioned)?

Since the flattened format requires multiple nodes, I/O ops like "create dir" are non-atomic. Otherwise it works just fine, but we want to drop flattening in the long run as it introduces a source for errors. And while confidentiality is the obvious security goal, we should not loose integrity and availability out of sight.

> What's your motivation for changing the format for Cryptomator? To get something faster than AES-CTR + HMAC?

We already use GCM for newer vaults (but we don't migrate existing data). This didn't require us to define a completely new format. Our motivation for a interoperable is threefold:

1. increase user acceptance, as our claim of not locking them in on a specific vendor is currently more of a theoretical claim. Of course everything is open source, but you can't _simply_ switch to a different product
2. focusing on the core product: We can tell people who ask for a great CLI app to use gocryptfs, users who want a sync tool instead of a VFS to use rclone or cryptsync and so on.
3. bundle knowledge and experience in one place. We have evolved our vault format over time due to problems we didn't think about in the previous version. For example, we didn't have file extensions at first, but it caused trouble during deletion in Google Drive. Experience is currently scattered among multiple projects and we want to end reinventing the wheel just because we can't swallow our pride when a competing product has already solved a certain problem. E.g. you are already supporting xattr encryption, we are still in the process of adding support.

Bonus motivation: Such a common standard should be initiated by FOSS projects, not GAFA. :wink:

**Comment by rfjakob** [comment]
Without directory flattening, implementation in gocryptfs is a lot easier 👍

Note that in gocryptfs, MKDIR is also non-atomic because of the gocryptfs.diriv file. I don't see a way around this, unfortunately.

**Reply by overheadhunter** [reply]
I see, but at least when the diriv file is written, you know for sure, everything is fine. And when you then move the dir, you no longer need to think about the iv file.

In a flat structure, you have to look in two places.

**Reply by rfjakob** [reply]
Maybe I'm missing something here, but moving a directory looks like a single rename in the ciphertext ( https://gist.github.com/rfjakob/393650b8dc5c9d473d01c5f290e9b460 )

**Reply by overheadhunter** [reply]
You're right, `move` is a bad example. But we still have more nodes than necessary, leaving us with more "partially synced" states.
Furthermore having the dir IV in place improves disaster recoverability and allows browser-based implementations that find all necessary info for decryption in a single dir listing.
Apps like our mobile apps or rclone that talk against storage backends will also profit from requiring fewer requests.

Side note: For best pagination support, assuming that sorting is done alphabetically, the diriv file should be prefixed with e.g. `_`.

**Reply by overheadhunter** [reply]
[Here is another example](https://github.com/cryptomator/cryptomator/issues/2713#issuecomment-1452200926), where a flattened structure is making things unnecessarily complicated: A subdirectory doesn't know which parent directory it belongs to. With a deep structure, you always know your complete path.

**Reply by overheadhunter** [reply]
And yet another reason to stick with deep structures: When applying file system permissions, you can safely rely on the ACL mechanism of the underlying file system without the need to manually update everything recursively.

We better not interfere with such capabilities of the native file system. If we decided against a 1:1 mapping of cleartext and ciphertext directory trees we would require complicated workarounds if we ever wanted to mimic such features.

Same rationale for sharing of subtrees.

**Reply by rfjakob** [reply]
Ack. Keeping the file hierarchy allows us to lean on the backing file system much more (e.g. ACLs)
